### PR TITLE
[PF-183] Billing info setup 

### PR DESCRIPTION
**Pull request by yonghaoy**

### CODE REVIEW

**Comment by yonghaoy** [comment]
Actually we need to enable billing api for the project we are using in test. Having a terraform for that...

File: src/main/java/bio/terra/rbs/service/resource/flight/GoogleProjectCreationFlight.java [hunk 1]
```diff
@@ -26,6 +29,10 @@ public GoogleProjectCreationFlight(FlightMap inputParameters, Object application
     addStep(new CreateResourceDbEntityStep(rbsDao));
     addStep(new GenerateProjectIdStep());
     addStep(new CreateProjectStep(rmCow, gcpProjectConfig), retryRule);
+    if (gcpProjectConfig.getBillingAccount() != null
+        && !gcpProjectConfig.getBillingAccount().isEmpty()) {
+      addStep(new SetBillingInfoStep(billingCow, gcpProjectConfig));
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/GoogleProjectCreationFlight.java:34]: maybe always add the step and check in the step if anything needs to be done based on the gcpProjectConfig? That might be a little simpler for us to debug Stairway things which care about the number of steps.

File: src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import static bio.terra.rbs.service.resource.FlightMapKeys.GOOGLE_PROJECT_ID;
+
+import bio.terra.cloudres.google.billing.CloudBillingClientCow;
+import bio.terra.rbs.generated.model.GcpProjectConfig;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import com.google.cloud.billing.v1.ProjectBillingInfo;
+
+/** Creates the basic GCP project. */
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java:12]: update

File: src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java [hunk 3]
```diff
@@ -50,16 +56,23 @@ public void testCreateGoogleProject() throws Exception {
...
+  private void assertProjectMatch(CloudResourceUid resourceUid, GcpProjectConfig gcpProjectConfig)
+      throws Exception {
     Project project =
         rmCow.projects().get(resourceUid.getGoogleProjectUid().getProjectId()).execute();
+    assertEquals(
+        gcpProjectConfig.getBillingAccount(),
+        billingCow
+            .getProjectBillingInfo("projects/" + project.getProjectId())
+            .getBillingAccountName());
     assertEquals("ACTIVE", project.getLifecycleState());
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java:76]: group the project get with the project lifecycle assertion - don't have the billing check in the middle.

File: src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java [hunk 4]
```diff
@@ -50,16 +56,23 @@ public void testCreateGoogleProject() throws Exception {
...
-  private void assertProjectMatch(CloudResourceUid resourceUid) throws Exception {
+  private void assertProjectMatch(CloudResourceUid resourceUid, GcpProjectConfig gcpProjectConfig)
+      throws Exception {
     Project project =
         rmCow.projects().get(resourceUid.getGoogleProjectUid().getProjectId()).execute();
+    assertEquals(
+        gcpProjectConfig.getBillingAccount(),
+        billingCow
+            .getProjectBillingInfo("projects/" + project.getProjectId())
+            .getBillingAccountName());
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java:75]: we're starting to get more detailed about checking that the project creation flight creates the project in the correct way. I think that kind of check is more appropriate for the `CreateProjectFlightIntegrationTest`. I also want a test that having a null billing account still allows a project to be created.

The tests in this file are just about the end2end, not the particulars how created projects.

**Review by wchamber — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/rbs/service/resource/flight/GoogleProjectCreationFlight.java (hunk 1 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/service/resource/flight/GoogleProjectCreationFlight.java:34]: done

File: src/main/java/bio/terra/rbs/service/pool/PoolWithResourceConfig.java [hunk 5]
```diff
@@ -7,9 +7,9 @@
 /** Represents a {@link PoolConfig} with corresponding {@link ResourceConfig}. */
 @AutoValue
 public abstract class PoolWithResourceConfig {
-  public abstract PoolConfig poolConfig();
+   abstract PoolConfig poolConfig();
 
-  public abstract ResourceConfig resourceConfig();
+   abstract ResourceConfig resourceConfig();
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/pool/PoolWithResourceConfig.java]: nit: it's funny that this class is public but it's methods are package private. I know they aren't currently used somewhere, but maybe keep them public for simplicity and so we don't have to change it later?

File: src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java (hunk 4 above)
Comment (yonghaoy) [reply src/test/java/bio/terra/rbs/integration/RbsIntegrationTest.java]: okay, moved to CreateProjectFlightIntegrationTest.

File: src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java [hunk 6]
```diff
@@ -21,6 +21,11 @@ public SetBillingInfoStep(CloudBillingClientCow billingCow, GcpProjectConfig gcp
 
   @Override
   public StepResult doStep(FlightContext flightContext) {
+    // Skip if billing account is not set.
+    if (gcpProjectConfig.getBillingAccount() != null
+        && !gcpProjectConfig.getBillingAccount().isEmpty()) {
+      return StepResult.getStepResultSuccess();
```
Comment (wchamber) [inline src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java:27]: uh, this looks backwards to me. Shouldn't we skip if the billing account is null or is an empty string?
How are the tests passing? Am I reading this wrong?

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java [hunk 7]
```diff
@@ -31,10 +34,35 @@
...
+    String flightId = manager.submitCreationFlight(pool).get();
+    blockUntilFlightComplete(flightId);
+    assertProjectCreated(pool);
+  }
+
+  @Test
+  public void testCreateGoogleProject_withBillingAccount() throws Exception {
+    // Basic GCP project with billing setup.
+    FlightManager manager = new FlightManager(flightSubmissionFactoryImpl, stairwayComponent);
+    Pool pool = preparePool(newBasicGcpConfig().billingAccount(BILLING_ACCOUNT_NAME));
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:60]: the basic project has a billing account, so we need a test for one without a billing account

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java [hunk 8]
```diff
@@ -129,22 +155,51 @@ private void blockUntilFlightComplete(String flightId)
...
+            .get(resource.cloudResourceUid().getGoogleProjectUid().getProjectId())
+            .execute();
+    assertEquals("ACTIVE", project.getLifecycleState());
+    if (pool.resourceConfig().getGcpProjectConfig().getBillingAccount() != null) {
+      assertEquals(
+          pool.resourceConfig().getGcpProjectConfig().getBillingAccount(),
+          billingCow
+              .getProjectBillingInfo("projects/" + project.getProjectId())
+              .getBillingAccountName());
+    }
```
Comment (wchamber) [inline src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:200]: nit: I think it'd be better to turn this into multiple asserts. I don't like the conditional asserting based on input parameters - that's too much logic for a test assertion.
maybe `void assertProjectExists(Resource resource)` and `assertBillingIs(Resource resource, String billingAccount)`?

**Review by wchamber — CHANGES_REQUESTED** [review]

File: src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java (hunk 6 above)
Comment (yonghaoy) [reply src/main/java/bio/terra/rbs/service/resource/flight/SetBillingInfoStep.java:27]: oh, right. Test does fails actually. Test passing now...

File: src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java (hunk 7 above)
Comment (yonghaoy) [reply src/test/java/bio/terra/rbs/integration/CreateProjectFlightIntegrationTest.java:60]: changed here to: Basic doesn't have billing account. We only set billing account in this test here.

File: terra-rbs-client/build.gradle [hunk 9]
```diff
@@ -0,0 +1,108 @@
+plugins {
```
Comment (wchamber) [inline terra-rbs-client/build.gradle:1]: uh, I think adding the `terra-rbs-client/build.gradle` should be a separate PR. Please remove. Careful with your commits and branches.

**Review by wchamber — APPROVED** [review]

File: terra-rbs-client/build.gradle (hunk 9 above)
Comment (yonghaoy) [reply terra-rbs-client/build.gradle]: ah, strange, it is in his PR. My local branchs got messed up somehow. Removed
